url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8/src/com/fuzzylite/imex/FclImporter.java

```diff
@@ -0,0 +1,613 @@
+            } else if ("ACT".equals(firstToken)) {
+                ruleBlock.setActivation(extractTNorm(line));
+            } else if ("ENABLED".equals(firstToken)) {
+                ruleBlock.setEnabled(extractEnabled(line));
+            } else if ("RULE".equals(firstToken)) {
+                int ruleStart = line.indexOf(':');
+                if (ruleStart < 0) {
+                    ruleStart = "RULE".length();
+                }
+                String rule = line.substring(ruleStart + 1).trim();
+                ruleBlock.addRule(Rule.parse(rule, engine));
+            } else {
+                throw new RuntimeException(String.format(
+                        "[syntax error] keyword <%s> not recognized in line %s",
+                        firstToken, line));
+            }
+        }
+    }
+
+    protected TNorm extractTNorm(String line) {
+        List<String> token = Op.split(line, ":");
+        if (token.size() != 2) {
+            throw new RuntimeException("[syntax error] "
+                    + "expected property of type (key : value) in line: " + line);
+        }
+        String name = token.get(1).trim();
+        String className = null;
+        if ("MIN".equals(name)) {
+            className = Minimum.class.getSimpleName();
+        } else if ("PROD".equals(name)) {
+            className = AlgebraicProduct.class.getSimpleName();
+        } else if ("BDIF".equals(name)) {
+            className = BoundedDifference.class.getSimpleName();
+        } else if ("DPROD".equals(name)) {
+            className = DrasticProduct.class.getSimpleName();
+        } else if ("EPROD".equals(name)) {
+            className = EinsteinProduct.class.getSimpleName();
+        } else if ("HPROD".equals(name)) {
+            className = HamacherProduct.class.getSimpleName();
+        }
+        return FactoryManager.instance().tnorm().createInstance(className);
+    }
+
+    protected SNorm extractSNorm(String line) {
+        List<String> token = Op.split(line, ":");
+        if (token.size() != 2) {
+            throw new RuntimeException("[syntax error] "
+                    + "expected property of type (key : value) in line: " + line);
+        }
+        String name = token.get(1).trim();
+        String className = name;
+        if ("MAX".equals(name)) {
+            className = Maximum.class.getSimpleName();
+        } else if ("ASUM".equals(name)) {
+            className = AlgebraicSum.class.getSimpleName();
+        } else if ("BSUM".equals(name)) {
+            className = BoundedSum.class.getSimpleName();
+        } else if ("NSUM".equals(name)) {
+            className = NormalizedSum.class.getSimpleName();
+        } else if ("DSUM".equals(name)) {
+            className = DrasticSum.class.getSimpleName();
+        } else if ("ESUM".equals(name)) {
+            className = EinsteinSum.class.getSimpleName();
+        } else if ("HSUM".equals(name)) {
+            className = HamacherSum.class.getSimpleName();
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 356:370, 473:487.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.